File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        return solve(prices);
    }
private:
    int solve(vector<int>& prices) {
        int bsp = -1 * prices[0], ssp = 0, csp = 0;
        
        for(int i = 1; i < prices.size(); ++i) {
            
            int nbsp = max(bsp, csp - prices[i]);
            int nssp = max(ssp, bsp + prices[i]);
            int ncsp = max(csp, ssp);
            
            bsp = nbsp;
            ssp = nssp;
            csp = ncsp;
        }
        return ssp;
    }
};

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        bsp = -1 * prices[0]
        ssp, csp = 0, 0
        
        for i in range(1, len(prices)):
            
            nbsp = max(bsp, csp - prices[i])
            nssp = max(ssp, bsp + prices[i])
            ncsp = max(csp, ssp)
            
            bsp, ssp, csp = nbsp, nssp, ncsp
        
        return ssp
        